finesse/gui/serial_device_panel.py

```diff
@@ -0,0 +1,53 @@
+"""Contains a panel which enables/disables child controls when device opens/closes."""
+from typing import Any
+
+from pubsub import pub
+from PySide6.QtWidgets import QGroupBox, QWidget
+
+
+class SerialDevicePanel(QGroupBox):
+    """A QGroupBox which enables/disables child controls when device opens/closes."""
+
+    def __init_subclass__(cls, **kwargs):
+        """Disable controls after construction."""
+        super().__init_subclass__(**kwargs)
+
+        def init_decorator(previous_init):
+            def new_init(self, *args, **kwargs):
+                previous_init(self, *args, **kwargs)
+                self.disable_controls()
+
+            return new_init
+
+        cls.__init__ = init_decorator(cls.__init__)
+
+    def __init__(self, name: str, title: str, *args: Any, **kwargs: Any) -> None:
+        """Create a new SerialDevicePanel.
+
+        Args:
+            name: The name of the device as used in pubsub messages
+            title: The title for the underlying QGroupBox
+            args: Extra arguments for the QGroupBox constructor
+            kwargs: Extra keyword arguments for the QGroupBox constructor
+        """
+        super().__init__(title, *args, **kwargs)
+
+        # Enable/disable controls on device connect/disconnect
+        pub.subscribe(self._on_device_opened, f"serial.{name}.opened")
+        pub.subscribe(self.disable_controls, f"serial.{name}.close")
+
+    def _on_device_opened(self) -> None:
+        self.enable_controls()
+
+    def enable_controls(self, enabled: bool = True) -> None:
+        """Enable the controls in this panel.
+
+        Args:
+            enabled: Whether to enable or disable the controls
+        """
+        for widget in self.findChildren(QWidget):
+            widget.setEnabled(enabled)
+
+    def disable_controls(self) -> None:
+        """Disable the controls in this panel."""
+        self.enable_controls(False)
```

Comment:
Why are these two public methods and why `disable_controls` is not just `_on_device_closed`, which is when it is used?

Reply:
They could be private, I guess. They're only called by `pubsub`.

I had it in mind that it was a better idea to name functions according to what they do, rather than when they're called, so that was the logic here, but I've definitely not been consistent with that throughout the project!

Reply:
IF that's the logic, then there should be no `_on_device_open`. Having said that, it is common to name callback functions based on the when they are called, so I don't think that's a bad approach, either. I've the feeling you've mix both approaches here. I would suggest to rename the functions to one of these two sets (names are in the order the functions are written):

- enable_controls, toggle_enable, disable_controls
- _on_device_open, toggle_enable, _on_device_close

I prefer the later set.

Reply:
Good shout. It is a bit inconsistent atm